File: DataStructures/Libraries/src/Graph.java
public class Graph
{
 private int adj[][];
 private int numberOfNodes;
 public int[] tmp;
 private int curr;
 public void addEdge(int node1,int node2)throws Exception
 {
  if(node1>=numberOfNodes||node2>=numberOfNodes)
   throw new Exception("Invalid node");

  adj[node1][node2]=1;
  adj[node1][node2]=1;
 }
 public void deleteEdge(int node1,int node2)throws Exception
 {
  if(node1>=numberOfNodes||node2>=numberOfNodes)
   throw new Exception("Invalid node");

  adj[node1][node2]=0;
  adj[node1][node2]=0;
 }
 public void dfsRecursive(int start, boolean visited[])
 {
  visited[start]=true;
  tmp[curr]=start;
  curr++;
  for(int v:adj[start])
   if(!visited[v])
    dfsRecursive(v,visited);
 }
 public int[] dfs() throws Exception
 {
  boolean[] visited = new boolean[numberOfNodes];
  curr=0;
  tmp=new int[numberOfNodes];
  for(int i=0;i<numberOfNodes;i++)
   if(!visited[i])
    dfsRecursive(i,visited);
  return tmp;
 }
 public void bfsRecursive(int start) throws Exception
 {
  Queue q= new Queue();
  q.enqueue(start);
  boolean visited[]=new boolean[numberOfNodes];
  visited[0]=true;
  while(q.size()>0)
  {
   int u=q.back();
   q.dequeue();
   tmp[curr]=u;
   curr++;
   for(int v:adj[u])
   {
    if(v==1&&!visited[v])
    {
     q.enqueue(v);
     visited[v]=true;
    } 
   }
  }
 }
 public int[] bfs() throws Exception
 {
  boolean visited[]=new boolean[numberOfNodes];
  tmp=new int[numberOfNodes];
  curr=0;
  for(int i=0;i<numberOfNodes;i++)
   if(!visited[i])
    bfsRecursive(i);
  return tmp;
 }


 public Graph(int numberOfNodes)
 {
  this.numberOfNodes=numberOfNodes;
  this.adj=new int[numberOfNodes][numberOfNodes];
 }
}

File: DSA/graph.cpp
#include <bits/stdc++.h>
using namespace std;

// printing function

void print(string message)
{
    cout << endl
         << message;
}

// Graph Class

class Graph
{
public:
    // Using adjency list for making this graph

    map<int, vector<int>> adj;
    int noOfNodes;

    // Constructor

    Graph(int nodes)
    {
        this->noOfNodes = nodes;
    }

    // For adding a edge

    void addEdge()
    {
        // Taking input from the user
        int node1;
        print("Enter the node1: ");
        cin >> node1;

        int node2;
        print("Enter the node2: ");
        cin >> node2;
        // Checking the nodes exist or not
        if (node1 < 0 || node1 >= this->noOfNodes)
        {
            print("Invalid Value.");
            return;
        }
        if (node2 < 0 || node2 >= this->noOfNodes)
        {
            print("Invalid Value.");
            return;
        }
        // If Node1 and Node2 are same do nothing. Treating a mistake by user
        if (node1 == node2)
        {
            return;
        }

        // check if edge already exist in node1 if not add it
        bool existInNode1 = false;
        for (auto i : adj[node1])
        {
            if (i == node2)
            {
                existInNode1 = false;
            }
        }

        if (!existInNode1)
        {
            adj[node1].push_back(node2);
        }
    }

    // To remove a edge

    void removeEdge()
    {
        // Taking input from the user

        int node1;
        print("Enter the node1: ");
        cin >> node1;

        int node2;
        print("Enter the node2: ");
        cin >> node2;
        // Checking the nodes exist or not
        if (node1 < 0 || node1 >= this->noOfNodes)
        {
            print("Invalid Value.");
            return;
        }
        if (node2 < 0 || node2 >= this->noOfNodes)
        {
            print("Invalid Value.");
            return;
        }
        // If Node1 and Node2 are same do nothing. Treating a mistake by user
        if (node1 == node2)
        {
            return;
        }

        // Removing the edge
        for (int i = 0; i < adj[node1].size(); i++)
        {
            if (adj[node1][i] == node2)
            {
                adj[node1].erase(adj[node1].begin() + i);
            }
        }

        return;
    }

    // Function to display the graph in adjacency list

    void display()
    {
        cout << endl;

        for (auto i : adj)
        {
            cout << i.first << " -> ";
            for (auto j : i.second)
            {
                cout << j << " ";
            }
            cout << endl;
        }
    }

    // BFS

    void bfs()
    {

        if (adj.size() == 0)
        {
            return;
        }

        print("BFS is -");

        // declaring an unordered map for marking visited
        unordered_map<int, bool> visited;
        for (auto i : adj)
        {

            // if not present in visited

            if (!visited[i.first])
            {

                queue<int> q;
                int front = i.first;
                q.push(front);
                visited[front] = 1;

                while (!q.empty())
                {
                    front = q.front();

                    // printing value

                    cout << front << " ";

                    // removing it from queue

                    q.pop();

                    // adding all the neighbouring edge in queue

                    for (auto i : adj[front])
                    {
                        if (!visited[i])
                        {
                            q.push(i);
                            visited[i] = 1;
                        }
                    }
                }
            }
        }
    }

    // DFS

    void recursiveFunctionForDfs(unordered_map<int, bool> &visited, int node)
    {
        cout << node << " ";
        visited[node] = 1;
        for (auto listItem : adj[node])
        {
            if (!visited[listItem])
            {
                recursiveFunctionForDfs(visited, listItem);
            }
        }
    }

    // dfs main function

    void dfs()
    {

        cout << endl
             << "DFS is -- ";
        unordered_map<int, bool> visited;

        for (auto i : adj)
        {
            if (!visited[i.first])
            {
                recursiveFunctionForDfs(visited, i.first);
            }
        }
    }
};

int main()
{

    int nodes;

    print("Enter the no. of nodes.");

    cin >> nodes;

    // Creating Graph Object

    Graph g(nodes);

    // while for menu --

    while (true)
    {
        print("Your Directed Graph -- ");
        g.display();

        cout << endl;

        print("Enter 1 for add edge operation");
        print("Enter 2 for remove edge operation");
        print("Enter 3 for BFS Traversal operation");
        print("Enter 4 for DFS Traversal operation");

        print("Enter 5 to exit");
        cout << endl;
        int input;
        cin >> input;
        int data;

        // On different input cases

        switch (input)
        {
        case 1:
            g.addEdge();
            break;
        case 2:
            g.removeEdge();
            break;
        case 3:
            g.bfs();
            break;
        case 4:
            g.dfs();
            break;

        default:
            return 0;
        }
    }
}